molgenis-data-semanticsearch/src/main/java/org/molgenis/data/semanticsearch/service/impl/SemanticSearchServiceImpl.java

```diff
@@ -63,22 +100,215 @@
 		return MetaUtils.toExistingAttributeMetaData(sourceEntityMetaData, attributeMetaDataEntities);
 	}
 
+	// TODO : remove the findAttributes method later on because of the duplicated code
+	public Iterable<ExplainedAttributeMetaData> explainAttributes(EntityMetaData sourceEntityMetaData,
+			EntityMetaData targetEntityMetaData, AttributeMetaData targetAttribute)
+	{
+		Iterable<String> attributeIdentifiers = semanticSearchServiceHelper
+				.getAttributeIdentifiers(sourceEntityMetaData);
+
+		QueryRule disMaxQueryRule = semanticSearchServiceHelper.createDisMaxQueryRuleForAttribute(targetEntityMetaData,
+				targetAttribute);
+
+		List<QueryRule> finalQueryRules = Lists.newArrayList(new QueryRule(AttributeMetaDataMetaData.IDENTIFIER,
+				Operator.IN, attributeIdentifiers));
+
+		if (disMaxQueryRule.getNestedRules().size() > 0)
+		{
+			finalQueryRules.addAll(Arrays.asList(new QueryRule(Operator.AND), disMaxQueryRule));
+		}
+
+		Iterable<Entity> attributeMetaDataEntities = dataService.findAll(AttributeMetaDataMetaData.ENTITY_NAME,
+				new QueryImpl(finalQueryRules));
+
+		Map<String, String> collectExpanedQueryMap = semanticSearchServiceHelper.collectExpanedQueryMap(
+				targetEntityMetaData, targetAttribute);
+
+		// Because the explain-API can be computationally expensive, therefore we limit the explanation to only top 10
+		// attributes
+		List<ExplainedAttributeMetaData> explainedAttributes = new ArrayList<ExplainedAttributeMetaData>();
+		int count = 0;
+		for (Entity attributeEntity : attributeMetaDataEntities)
+		{
+			ExplainedAttributeMetaData explainedAttributeMetaData;
+			if (count < 10)
+			{
+				explainedAttributeMetaData = convertAttributeEntityToExplainedAttribute(attributeEntity,
+						sourceEntityMetaData, collectExpanedQueryMap, finalQueryRules);
+			}
+			else
+			{
+				AttributeMetaData attribute = sourceEntityMetaData.getAttribute(attributeEntity
+						.getString(AttributeMetaDataMetaData.NAME));
+				explainedAttributeMetaData = new ExplainedAttributeMetaData(attribute, Collections.emptySet());
+			}
+			explainedAttributes.add(explainedAttributeMetaData);
+			count++;
+		}
+
+		return explainedAttributes;
+	}
+
+	/**
+	 * A helper function to explain each of the matched attribute by explain-api
```

Comment:
A helper function to explain each of the matched attributes returned by the explain-API